### ISSUE TITLE
How to use the library as binary classifier for complex object

### ISSUE THREAD

**Issue by taglia3** [collinhundley/Swift-AI#65]
Hi,
your library looks pretty. I would like to know if I can use it as **classifier** for data.

Example:
### Given a _human_ object classified it as good or bad.

Consider this scenario:

Given a _Human_ struct:

```
struct Human {
    var name: String
    var title: String
    var job: Job
    var residence: Location
    var isMarried: Bool
}
```
where:

```
struct Job {
    var salary: Double
}

struct Location {
    var address: String
}
```

and this extract of training data:
```
let human1 = Human(name: "Jonh", title: "Eng.", job: Job(salary: 2000), residence: Location(address: "Street 5"), isMarried: true)
let human2 = Human(name: "Albert", title: "Mr.", job: Job(salary: 0), residence: Location(address: "Street 2"), isMarried: false)
let human3 = Human(name: "Wayne", title: "Mrs.", job: Job(salary: 1000), residence: Location(address: "Street 4"), isMarried: true)
let human4 = Human(name: "Luke", title: "Doctor", job: Job(salary: 4000), residence: Location(address: "Street 5"), isMarried: false)

where human1 -> good
where human2 -> good
where human3-> bad
where human4 -> bad
```

How can I use your library as classifier for a new human object like this one? 

```
let humanToClassified = Human(name: "Dart", title: "Developer", job: Job(salary: 2000), residence: Location(address: "Street 3"), isMarried: true)
```

Is humanToClassified good or bad ?

**Comment by collinhundley** [comment]
Hi @taglia3, neural networks like this one generally work well as classifiers.

What I'm confused about is how you are determining whether a `Human` is `good` or `bad` - can you provide some more context?

Your data implies that the "goodness" of a `Human` can be determined based on their name, title, salary, address and marital status. Of course in the real world this isn't true, but maybe in your application it can work. We'll need some more information to know for sure.

**Comment by taglia3** [comment]
Hi @collinhundley,
thanks for the quick response!
I know that the example isn't very explicit and consistent with reality, but I just wanted to figure out how I should handle a situation like this one by using your library.

I would like to understand how the library learn rules based on custom objects as in example, and apply them to objects to be classified.

I posted a simple example, but you can very well change it as you want.

The main question are:
How I could set the stage for traning with the know set of Human objects?
How could I set the classification stage?

Thank you

**Comment by collinhundley** [comment]
@taglia3, the biggest hurdle for something like this is figuring out how to encode your data.

Your example contains a lot of text data, which makes it a poor fit for a simple 3-layer network (unless there were a limited number of names/titles/addresses to choose from - in which case, things get easier). If you *really* need to make inferences based on textual data, you'll do better with a Recurrent Neural Network (coming soon!).

Don't worry about the "sub-objects", we can just inline them directly into the top-level `Human` object. For educational purposes, let's simplify our model, get rid of the `Strings`, and add some new numerical properties:

```swift
struct Human {
    age: Double
    salary: Double
    distanceFromCity: Double
    numChildren: Int
    isMarried: Int  // still a Bool, but encoded as 0/1
}
```

Now we've got something we can work with. At this point, you'll want to normalize each property to have a mean of `0` and a standard deviation of `1`. [This is a good reference for normalization](https://visualstudiomagazine.com/articles/2014/01/01/how-to-standardize-data-for-neural-networks.aspx).

Then we construct a neural network with 5 inputs, ~3 hidden nodes, and 2 outputs.  Instructions for training and inference can be found in the docs and lots of places online :)

Does this help?

**Comment by collinhundley** [comment]
Closing this.  Hope it helped 👍🏼
